File: Array/medium/longest_sequence_in_array.cpp
//
// Created by santosh on 12/21/25.
//

#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:

    int getLongestSequence(vector<int> &nums)
    {
        unordered_set<int> uniqueElements;
        int longestSequence=INT_MIN;
        int count=0;

        for(auto num:nums){
            uniqueElements.insert(num);
        }

        for(auto num : uniqueElements){
            if(uniqueElements.find(num-1)==uniqueElements.end()){
                count=1;
                int x=num;
                while(uniqueElements.find(x+1)!=uniqueElements.end()){
                    count++;
                    x+=1;
                }
                longestSequence= max(longestSequence, count);
            }
        }

        return longestSequence;

    }

};


int main()
{
    Solution s;
    vector<int> nums={100,4,200,1,3,2};
    cout<< s.getLongestSequence(nums);
}

File: Array/medium/longest_sequence_in_array.py
# Longest consecutive sequence in an array (LeetCode 128)
# Created by santosh

class Solution:
    def get_longest_sequence(self, nums: list[int]) -> int:
        """
        HashSet approach: only start counting from sequence beginners.
        Time: O(n), Space: O(n)
        """
        unique_elements = set(nums)
        longest_sequence = float('-inf')

        for num in unique_elements:
            # Only start a sequence from its beginning
            if (num - 1) not in unique_elements:
                count = 1
                x = num
                while (x + 1) in unique_elements:
                    count += 1
                    x += 1
                longest_sequence = max(longest_sequence, count)

        return longest_sequence


if __name__ == "__main__":
    s = Solution()
    nums = [100, 4, 200, 1, 3, 2]
    print(s.get_longest_sequence(nums))  # Expected: 4 (1,2,3,4)
